### PR TITLE
fix(core): correct keyed segment path representation and insert opera…

### PR DESCRIPTION
**Pull request by ricokahler**
### Description

This PR introduces several important fixes and improvements to the patch operations implementation in `packages/core/src/document/patchOperations.ts`:

**Key Changes:**
1. **Fixed keyed segment path representation**: When matching array elements by `_key`, the returned path now correctly uses the keyed segment format (`{_key: "value"}`) instead of the numeric index. This ensures path consistency and proper semantic meaning.

2. **Simplified and fixed range matching**: Replaced the manual for-loop in range matching with a more concise `flatMap` implementation that correctly calculates indices.

3. **Completely rewrote the `insert` function**: 
   - Simplified the complex grouping logic with a more straightforward approach
   - Fixed behavior to match content-lake expectations for `before`, `after`, and `replace` operations
   - Improved handling of edge cases and array index calculations
   - Better support for keyed segments in insert operations

4. **Code cleanup**: Removed unused code including the `operations` array, `Operation` type, and `isNonNullable` helper function.

5. **Fixed typo**: Corrected `SupportPatchOperation` to `SupportedPatchOperation` in `processMutations.ts`.

These changes address inconsistencies in how patch operations handle array elements with keys and improve the overall reliability of the patch system.

### What to review

Reviewers should focus on:

1. **Path representation consistency**: Verify that `jsonMatch` now correctly returns keyed segments in paths rather than numeric indices when matching by `_key`
2. **Insert operation behavior**: Test the `insert` function with various scenarios including:
   - `before`, `after`, and `replace` operations
   - Operations on keyed array elements
   - Edge cases with negative indices and out-of-bounds scenarios
3. **Range matching**: Ensure range operations like `items[1:3]` work correctly with the new implementation
4. **Test coverage**: Review the updated test case that now expects keyed segment paths
5. **Backward compatibility**: Verify that existing patch operations continue to work as expected

The changes primarily affect the `packages/core` patch operations and any code that relies on path representation from `jsonMatch`.

### Testing

- Updated the existing test case for keyed segment matching to reflect the correct path format
- The changes maintain the existing test suite while fixing the underlying implementation
- Manual testing should be performed on various patch operation scenarios, particularly those involving arrays with keyed elements

Additional testing recommendations:
- Test insert operations with mixed keyed and indexed array elements
- Verify range operations work correctly across different array sizes
- Test edge cases with negative indices and out-of-bounds scenarios

### Fun gif

Fixing things like a pro

### CODE REVIEW

File: packages/core/src/document/patchOperations.ts [hunk 1]
```diff
@@ -579,105 +568,135 @@ type Operation = (typeof operations)[number]
...
-  // For each group, update the parent array using setDeep.
-  const result = groupEntries.reduce<unknown>((acc, {array, indexes, pathToArray}) => {
     switch (operation) {
+      case 'replace': {
+        const indexesToRemove = new Set<number>()
+        let position = Infinity
+
+        for (const itemMatch of jsonMatch(arr, positionPath)) {
+          // there should only be one path segment for an insert patch, invalid otherwise
+          if (itemMatch.path.length !== 1) continue
```
[the remark this answers is not in the document]
Comment (ricokahler) [reply packages/core/src/document/patchOperations.ts:617]: I think I was testing this out this case in content-lake and it seems like it lets it go through but it results in a no-op so not throwing and skipping seems like the right behavior
